### PR TITLE
[fix] sliding window attention mask

### LINKED ISSUE #38002: [bug] `use_sliding_window` doesn't work as expected
### System Info

* transformer: main
* pytorch, cuda: anyversion

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
from transformers import AutoModelForCausalLM
import numpy as np
import torch


MODEL_NAME = "deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B"
model = AutoModelForCausalLM.from_pretrained(MODEL_NAME)

input_ids = torch.randint(0, 100, (1, 8192))

with torch.no_grad():
    output_raw = model(input_ids)

# correct situaion: should be the same as the original model since use_sliding_window is False
model_no_sliding = AutoModelForCausalLM.from_pretrained(MODEL_NAME, sliding_window=None)

with torch.no_grad():
    output_non_sliding = model_no_sliding(input_ids)


np.testing.assert_allclose(output_raw.logits[:, :4096], output_non_sliding.logits[:, :4096])

# wrong: the logits are unexpectedly different with sliding_window=4096
np.testing.assert_allclose(output_raw.logits[:, 4096:], output_non_sliding.logits[:, 4096:])
```

### Expected behavior

# description
What is expected:   

```
  "sliding_window": 4096,
  "use_sliding_window": false,
```
`use_sliding_window` is set as false in `deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B` [here](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B/blob/main/config.json#L20). We do expect sliding window is disabled. In other words, we should expect the same results even with different `sliding_window`.

However, the results are different in the repro script.

# Root cause
Attention Mask is changed according to `sliding_window` without respect on `use_sliding_window`.

https://github.com/huggingface/transformers/blob/3c0796aaeac722152836ea063d542a4c628a75be/src/transformers/models/qwen2/modeling_qwen2.py#L708-L715

If we add some printing under this conditional block, we can clearly see attention mask is changed even with `use_sliding_window=false`

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/38002. Sliding mask should not be applied when `config.use_sliding_window = False`

Added a small test as well

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -4308,6 +4308,44 @@ def _get_custom_4d_mask_test_data(self):
 
         return input_ids, position_ids, input_ids_shared_prefix, mask_shared_prefix, position_ids_shared_prefix
 
+    def test_sliding_window_mask(self):
```
Comment (gante) [inline tests/test_modeling_common.py:4311]:
```suggestion
    def test_sliding_window_mask(self):
        """ Tests that we can control the sliding window attention behavior of a model. """
```
(So we can quickly understand what this test should be doing. Useful when refactoring multiple tests.)

**Review by gante — APPROVED** [review]
LGTM, and thank you for adding a test 🙏

**Review by ArthurZucker — APPROVED** [review]
Hahah okay I think I reviewed a similar PR recently, saying we should isolate to models that can have `sliding_window` not defined in the config, but this is better I think. 
Mit supprised that moshi or mimi are affected as I did not know they had sliding but they prob do!
